feat: add MaterialBindingAPIChecker

beersandrew opened the pull request:
### Description of Change(s)
- Added all necessary validation pieces including plugInfo.json, valdator tokens, tests, and validator code itself for the MaterialBindingAPIChecker

### Fixes Issue(s)
- https://github.com/orgs/PixarAnimationStudios/projects/3?pane=issue&itemId=65425035

<!--
Please follow the Contributing and Building guidelines to run tests against your
change. Place an X in the box if tests are run and are all tests passing.
-->
- [X] I have verified that all unit tests pass with the proposed changes
<!-- 
Place an X in the box if you have submitted a signed Contributor License Agreement.
A signed CLA must be received before pull requests can be merged.
For instructions, see: http://openusd.org/release/contributing_to_usd.html
-->
- [X] I have submitted a signed Contributor License Agreement

beersandrew committed:
feat: add MaterialBindingAPIChecker

- adding all necessary validation pieces including plugInfo.json, valdator tokens, tests, and validator code itself

jesschimein commented:
Filed as internal issue #USD-9833

jesschimein commented:
/AzurePipelines run

dgovil commented on pxr/usd/usdShade/validators.cpp:
```diff
@@ -192,13 +193,45 @@ _ShaderPropertyTypeConformance(const UsdPrim &usdPrim)
+{
+    UsdValidationErrorVector errors;
+
+    if (!usdPrim.CanApplyAPI<UsdShadeMaterialBindingAPI>()){
```
Question for the group: I know the current check doesn't check for it, but should the test flag a warning (not an error) when the MaterialBindingAPI is applied to a prim that it can't be applied to?

tallytalwar replied:
What is an example of such an application @dgovil? I thought MaterialBindingAPI can be applied to any prim which wants to provide a material:binding relationship.

dgovil replied:
Oh is MaterialBindingAPI unbounded? I thought it required at least being a gprim?

tallytalwar replied:
I think a prim's ancestor also particate in determining what material is bound: https://github.com/PixarAnimationStudios/OpenUSD/blob/release/pxr/usd/usdShade/materialBindingAPI.cpp#L718, so there is no restriction on a type of prim which can't have MaterialBindingAPI applied. Also if there was such a restriction, the schema should be tightened by having "CanOnlyApplyTo" in schema.usda.

dgovil commented:
Also LGTM! Great work :-)

tallytalwar commented on pxr/usd/usdShade/plugInfo.json:
```diff
@@ -133,7 +133,10 @@
                         ]
-                    }, 
+                    },
+                    "MaterialBindingApiChecker": {
```
Checker or not to Checker?
MaterialBindingAPIApplicationValidator, MaterialBindingAPIApplicationCompliance?

tallytalwar commented on pxr/usd/usdShade/plugInfo.json:
```diff
@@ -133,7 +133,10 @@
-                    }, 
+                    },
+                    "MaterialBindingApiChecker": {
+                        "doc": "Verify a prim has the MaterialBindingAPI applied if it has a material relationship binding."
```
material binding relationship.

tallytalwar commented on pxr/usd/usdShade/testenv/testUsdShadeValidators.cpp:
```diff
@@ -118,11 +122,50 @@ TestUsdShadeShaderPropertyCompliance()
 int
 main()
 {
     TestUsdShadeValidators();
```
Can you also include this validator to the list above for TestUsdShadeValidators?

tallytalwar commented on pxr/usd/usdShade/validators.cpp:
```diff
@@ -192,13 +193,45 @@ _ShaderPropertyTypeConformance(const UsdPrim &usdPrim)
+
+    UsdShadeMaterialBindingAPI materialBindingAPI(usdPrim);
+
+    if (materialBindingAPI.GetDirectBinding().IsBound() && !usdPrim.HasAPI<UsdShadeMaterialBindingAPI>()) {
```
Hmm.. this is only checking for direct binding, but a prim could also provide a collection based binding rel, which I think it will not cover.

Might be simpler to do the exact same thing which is being done in the python code.. that is, get all relationships in "materialBinding" namespace / startswith, if exists and API not applied report error.

tallytalwar reviewed (commented):
Left some notes for you to consider @beersandrew. Thanks

beersandrew committed:
fix: addressed PR comments

- updated plugInfo description
- updated name to MaterialBindingApiAppliedValidator
- updated how material binding relationships are found to be how the python compliance checker was checking it
-

jesschimein commented:
/AzurePipelines run

tallytalwar commented on pxr/usd/usdShade/validators.cpp:
```diff
@@ -192,13 +192,48 @@ _ShaderPropertyTypeConformance(const UsdPrim &usdPrim)
     return errors;
 }
 
+static bool _HasMaterialBindingRelationship(const UsdPrim& usdPrim) {
```
Might be better to make this a lambda within `_MaterialBindingApiAppliedValidator` method?

And also make the `materialBindingString` a `static const`, or its totally fine if you do `UsdShadeTokens->materialBinding.GetString()` in place.

tallytalwar reviewed (approved):
Just 1 cleanup comment. Otherwise lgtm. Thanks @beersandrew

beersandrew committed:
fix: addressed PR comments

- converted hasMaterialBindingRelationship function into a lambda
- changed materialBindingString to be static

jesschimein commented:
/AzurePipelines run

pablode commented on pxr/usd/usdShade/testenv/testUsdShadeValidators.cpp:
```diff
@@ -118,11 +123,50 @@ TestUsdShadeShaderPropertyCompliance()
+
+    UsdValidationErrorVector errors = validator->Validate(usdPrim);
+
+    auto site = errors[0].GetSites()[0];
```
(leftover line / before size check)

beersandrew replied:
fixed, thanks!

beersandrew committed:
fix: remove unused line

jesschimein commented:
/AzurePipelines run
